### ISSUE TITLE
fs.copyFile fails with permission denied when source is read-only

### ISSUE THREAD

**Issue by RaisinTen** [libuv/libuv#3117]
See: https://github.com/nodejs/node/issues/37284
cc @kakaroto

**Comment by bnoordhuis** [comment]
> `Platform: Linux 4.19.0-11-amd64 #1 SMP Debian 4.19.146-1 (2020-09-17) x86_64 Linux`

See torvalds/linux@6f9718fe41c3a47e4362bddf145e2db6ad7d8e87 - it's a cephfs bug that was fixed (well, mitigated) in Linux 4.20. There's no workaround but I'll open a pull request with a fallback.

**Comment by kakaroto** [comment]
Nice find! Can't wait to test it out. Not sure why Node 12.18.4 works (it might be something else behind this, since NFS also seemed affected by the bug I found), but I have high hopes that this would fix it too. Will give that a try as soon as it's available in an update for me to test with.
Thanks for finding/implementing the workaround so quickly 👍

**Comment by kakaroto** [comment]
I see that this fix was released as part of libuv 1.42.0 which was included in node 16.7.0 released last week.
I've tested my previous test case with node 16.7.0 and unfortunately the issue is still there.

Note that I was also able to reproduce a similar problem with a `cifs` mount (which also affects 12.18.x) and which also still happens in 16.7.0. Hopefully, with `cifs` (SMB mount), it should make it much easier to test and reproduce the issue on dev environments.

`docker run --rm -it -v /cifs-mount/tmp:/mount node:12.19.0-alpine /bin/sh`

```
su - node
echo foo > /tmp/foo
chmod -w /tmp/foo
cat <<EOF > test.js
  const fs = require("fs");
  fs.copyFileSync("/tmp/foo", "/mount/bar");
EOF

rm -f /mount/bar
node test.js
exit
exit
```

(Note that running the same commands as above but without the `chmod -w` on the input file does not result in any errors.)

The difference between `cifs` and `cephfs` mounts in the behavior is that cephfs doesn't have the issue on node 12.18.4 (libuv 1.38.0), but has the problem on node 12.19.0 (libuv 1.39.0), while cifs has the problem on both versions. The other difference is with cifs, the target file actually gets created (but empty), despite the fact that it throws the EPERM error, while on cephfs, it will not create the target file at all.
Note that I tested/confirmed the cephfs behavior on Linux 4.19.0-13 (`Linux 4.19.0-13-cloud-amd64 #1 SMP Debian 4.19.160-2 (2020-11-28) x86_64 GNU/Linux`) while I tested/confirmed the cifs behavior on Linux 5.0.16-100 (`Linux 5.0.16-100.fc28.x86_64 #1 SMP Tue May 14 18:22:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`)
